src/Linters/UnusedVariableLinter.hack

```diff
@@ -0,0 +1,289 @@
+      $v ==> ($v->getExpression() as VariableToken)->getText(),
+    );
+    $ret['used'] = Keyset\map(
+      $ret['used'],
+      $v ==> ($v->getExpression() as VariableToken)->getText(),
+    );
+
+    return $ret;
+  }
+
+  /**
+   * Return whether `$var` is being assigned in the scope of `$parents`.
+   */
+  private function isVariableAssignment(
+    VariableExpression $var,
+    vec<Node> $parents,
+  ): bool {
+    foreach (Vec\reverse($parents) as $parent) {
+      if ($parent === $var) {
+        continue;
+      }
+
+      // Too early to tell
+      if ($parent is ListItem<_> || $parent is NodeList<_>) {
+        continue;
+      }
+
+      // SubscriptExpression
+      // $a[$b]
+      // ^^ ^^
+      // |  +-- index
+      // +-- receiver
+      //
+      // A variable in the index of a subscript expression is never an assignment
+      // A variable that is the receiver of a subscript expression may be an assignment or use; continue until a BinaryExpression is found.
+      //
+      // $a[$b] = '123'; // assignment of $a, use of $b
+      // f($a[$b]);      // use of $a
+      if ($parent is SubscriptExpression) {
+        $index = $parent->getIndex();
+        if ($index !== null && $index->isAncestorOf($var)) {
+          return false;
+        } else {
+          continue;
+        }
+      }
+
+      if (
+        $parent is BinaryExpression &&
+        $this->isAssignmentExpression($parent) &&
+        $parent->getLeftOperand()->isAncestorOf($var)
+      ) {
+        return true;
+      }
+
+      if (
+        $parent is ForeachStatement &&
+        (($parent->getValue() === $var) || ($parent->getKey() === $var))
+      ) {
+        return true;
+      }
+
+      if (
+        $parent is ListExpression &&
+        (bool)$parent->getMembers()?->isAncestorOf($var)
+      ) {
+        return true;
+      }
+
+      break;
+    }
+
+    return false;
+  }
+
+  /**
+   * These are all the return types of BinaryExpression::getOperator
+   * that contain "Equal" and are not comparison operators.
+   */
+  private function isAssignmentExpression(BinaryExpression $expr): bool {
+    $op = $expr->getOperator();
+    return (
+      $op is AmpersandEqualToken ||
+      $op is BarEqualToken ||
+      $op is CaratEqualToken ||
+      $op is DotEqualToken ||
+      $op is EqualToken ||
+      $op is GreaterThanGreaterThanEqualToken ||
+      $op is LessThanLessThanEqualToken ||
+      $op is MinusEqualToken ||
+      $op is PercentEqualToken ||
+      $op is PlusEqualToken ||
+      $op is QuestionQuestionEqualToken ||
+      $op is SlashEqualToken ||
+      $op is StarEqualToken ||
+      $op is StarStarEqualToken
+    );
+  }
+
+  public function getFixedNode(VariableExpression $node): VariableExpression {
+    $expr = $node->getExpression();
+    if (!$expr is VariableToken) {
+      return $node;
+    }
+    return $node->withExpression(
+      $expr->withText('$_'.Str\strip_prefix($expr->getText(), '$')),
+    );
+  }
```

Comment:
Wouldn't it be a better fix to remove the whole assignment? (except for special cases like `list($_unused, $used) = ...`)

I'm slightly concerned that this suggest what is arguably the wrong fix in most cases, but we can discuss that later in a separate issue/PR.

I imagine that correctly removing the assignment could be much harder.